content/2_Bronze/Intro_DS.mdx

```diff
@@ -159,8 +159,80 @@ dynamic arrays (e.g. `vector<vector<int>>`) static arrays of dynamic arrays
 (e.g. `array<vector<int>,5>`), dynamic arrays of static arrays (e.g.
 `vector<array<int,5>>`), and so on.
 
-<!-- Usually, the choice between a static array and a dynamic array is just personal preference. -->
+## Doubly-Linked Lists
+	
+Another type of datastructure is a "doubly-linked list." They are implemented by "linking" or pointing each element to the element before and after it. We recommend you look at [this Wikipedia article](https://en.wikipedia.org/wiki/Doubly_linked_list) for more information. These support $\mathcal{O}(1)$ insertion and removal if you have access to the iterator of the element to insert before or remove beforehand.
+	
+<Warning>
+	Doubly-Linked lists cannot be accessed by indexes. This is a major reason why they are not used, as opposed to vectors or regular arrays.
+</Warning>
+	
+The C++ object for a doubly-linked list is the `list` object.
+
+```cpp
+list<int> li
+```
+
+Alternatively, you could implement such a list on your own:
+	
+```cpp
+struct Node {
+    int value = NULL;
+    Node *prev = NULL;
+    Node *next = NULL;
+};
+
+void insert(Node* node, int value) {
+    struct Node* new_node = new Node;
+    new_node->value = value;
+    new_node->next = node->next;
+    new_node->prev = node;
+    node->next = new_node;
+    if (new_node->next != NULL) new_node->next->prev = new_node;
+}
+
+void remove(Node* node) {
+    if (node->prev != NULL) node->prev->next = node->next;
+    if (node->next != NULL) node->next->prev = node->prev;
+}
+```
+A sample implementation of this is provided below:
+
+```cpp
+int main() {
+	Node* start = new Node;
+	    start->value = 5;
+	
+	    insert(start, 6);
+	    Node* next = start->next;
+	
+	    insert(next, 7);
+	    next = next->next;
+	
+	    insert(next, 8);
+	
+	    remove(next);
+	
+	    Node* cur = start;
+	    while (cur != NULL) {
+			cout << cur->value << endl;
+			cur = cur->next;
+	    }
+
+	    // Infinite Loop
+	    next = next->next;
+	    next->next = start;
+	
+	    cur = start;
+	    while (cur != NULL) {
+			cout << cur->value << endl;
+			cur = cur->next;
+	    }
+}
+```
 
+Notice how you can create infinite loops with doubly-linked lists. Doubly-linked lists can be used to represent functional graphs. You can read more about functional graphs [here.](https://usaco.guide/silver/func-graphs)
```

Comment:
I still don't know what this means. presumably, if you mess up the pointers you'll get an infinite loop when you try to traverse such a list, but why would you purposely do this.

Reply:
Cycle Detection Algorithms for finding Cycles in functional graphs

Reply:
why wouldn't you use an array to represent a functional graph??

Reply:
I guess you could. But this makes more intuitive sense. And if there is a cycle, you would never know with an array.

Reply:
What do you mean by intuitive sense? Using a doubly-linked list for a functional graph makes no sense whatsoever, and using a singly linked list is equivalent to an array??

> And if there is a cycle, you would never know with an array.

???

Reply:
If there is a cycle in the graph, an array would not be able to represent that. The values might be duplicated, but that doesn't imply a cycle. For example, an array would represent:

1-2-3-1

While a doubly-linked list would represent

a(1)-b(2)-c(3)-a(1)

However, the doubly linked list could also be

a(1)-b(2)-c(3)-d(1)

and the array would be the same. One has a cycle, the other doesnt

Reply:
I meant successor graphs, not functional graphs. And sometimes it can be useful to have a reversed edge, right?

Reply:
er, I'm assuming you have a functional graph where vertices are represented by numbers, not letters. Also, not all functional graphs are cycles.

Reply:
successor graphs are the same thing as functional graphs

Reply:
Hm I think I understand what you are saying. An array might work, but I still feel like we should keep this in.

Reply:
What do you mean by "might?" It *will* work. 

I still don't quite understand what your representation is. Could you provide an example problem and code where using a linked list to represent a functional graph would be helpful?

Reply:
See the problem above. It can be used to solve the problem.

Reply:
code?

Reply:
what does it have to do w/ functional graphs?

Reply:
Oh wait yeah nvm sorry. I don't have one then.